src/QirTools/pyqir/qirlib/src/emit/mod.rs

```diff
@@ -0,0 +1,276 @@
+            qubits.insert(indexed_name, value);
+        }
+        qubits
+    }
+
+    fn write_registers<'ctx>(
+        model: &SemanticModel,
+        context: &Context<'ctx>,
+    ) -> BTreeMap<String, (BasicValueEnum<'ctx>, Option<u64>)> {
+        let mut registers = BTreeMap::new();
+        let number_of_registers = model.registers.len() as u64;
+        if number_of_registers > 0 {
+            let results =
+                qir::array1d::emit_array_allocate1d(&context, 8, number_of_registers, "results");
+            registers.insert(String::from("results"), (results, None));
+            let mut sub_results = vec![];
+            for reg in model.registers.iter() {
+                let (sub_result, entries) =
+                    qir::array1d::emit_array_1d(context, reg.name.as_str(), reg.size.clone());
+                sub_results.push(sub_result);
+                registers.insert(reg.name.clone(), (sub_result, None));
+                for (index, _) in entries {
+                    registers.insert(format!("{}{}", reg.name, index), (sub_result, Some(index))); 
+                }
+            }
+            qir::array1d::set_elements(&context, &results, sub_results, "results");
+            registers
+        } else {
+            let results = qir::array1d::emit_empty_result_array_allocate1d(&context, "results");
+            registers.insert(String::from("results"), (results, None));
+            registers
+        }
+    }
+
+    fn write_instructions<'ctx>(
+        model: &SemanticModel,
+        context: &Context<'ctx>,
+        qubits: &BTreeMap<String, BasicValueEnum<'ctx>>,
+        registers: &BTreeMap<String, (BasicValueEnum<'ctx>, Option<u64>)>,
+    ) {
+        for inst in model.instructions.iter() {
+            qir::instructions::emit(context, inst, qubits, registers);
+        }
+    }
+}
+
+pub struct Context<'ctx> {
+    pub(crate) context: &'ctx inkwell::context::Context,
+    pub(crate) module: inkwell::module::Module<'ctx>,
+    pub(crate) builder: inkwell::builder::Builder<'ctx>,
+    pub(crate) types: Types<'ctx>,
+    pub(crate) runtime_library: RuntimeLibrary<'ctx>,
+    pub(crate) intrinsics: Intrinsics<'ctx>,
+    pub(crate) constants: Constants<'ctx>,
+}
+impl<'ctx> Context<'ctx> {
+    pub fn new(context: &'ctx inkwell::context::Context, name: &'ctx str) -> Self {
+        let builder = context.create_builder();
+
+        let module = qir::load_module_from_bitcode_file(&context, name);
+
+        let types = Types::new(&context, &module);
+        let runtime_library = RuntimeLibrary::new(&module);
+        let intrinsics = Intrinsics::new(&module);
+        let constants = Constants::new(&module, &types);
+        Context {
+            builder: builder,
+            module: module,
+            types: types,
+            context: context,
+            runtime_library: runtime_library,
+            intrinsics: intrinsics,
+            constants: constants,
```

Comment:
Since each local variable here has the same name as the struct field being set, you can use the short-hand notation provided for struct initialization to reduce duplication. I believe this is also suggested by `cargo clippy` as well.

```suggestion
            builder,
            module,
            types,
            context,
            runtime_library,
            intrinsics,
            constants,
```